internal/processing/processors/authorization_policy_processor.go

```diff
@@ -34,34 +40,101 @@ func (r AuthorizationPolicyProcessor) EvaluateReconciliation(ctx context.Context
-	return r.Creator.Create(api)
+func (r AuthorizationPolicyProcessor) getDesiredState(api *gatewayv1beta1.APIRule) (map[string][]*securityv1beta1.AuthorizationPolicy, error) {
+	aps, err := r.Creator.Create(api)
+	if err != nil {
+		return nil, err
+	}
+	return aps, nil
 }
 
-func (r AuthorizationPolicyProcessor) getActualState(ctx context.Context, client ctrlclient.Client, api *gatewayv1beta1.APIRule) ([]*securityv1beta1.AuthorizationPolicy, error) {
+func (r AuthorizationPolicyProcessor) getActualState(ctx context.Context, client ctrlclient.Client, api *gatewayv1beta1.APIRule) (map[string]map[string]*securityv1beta1.AuthorizationPolicy, error) {
 	labels := processing.GetOwnerLabels(api)
 
 	var apList securityv1beta1.AuthorizationPolicyList
 	if err := client.List(ctx, &apList, ctrlclient.MatchingLabels(labels)); err != nil {
 		return nil, err
 	}
 
-	authorizationPolicies := append([]*securityv1beta1.AuthorizationPolicy{}, apList.Items...)
+	authorizationPolicies := make(map[string]map[string]*securityv1beta1.AuthorizationPolicy)
+	for i, ap := range apList.Items {
+		index, ok := ap.Labels[processing.IndexLabelName]
+		if !ok {
+			// Store unindexed APs so they will be deleted/updated with index
+			index = "-" + strconv.Itoa(i)
+		}
+
+		if hash, ok := ap.Labels[processing.HashLabelName]; ok {
+			if authorizationPolicies[hash] == nil {
+				authorizationPolicies[hash] = make(map[string]*securityv1beta1.AuthorizationPolicy)
+			}
+
+			if _, ok := authorizationPolicies[hash][index]; ok {
+				authorizationPolicies[hash][index+":"+strconv.Itoa(i)] = ap
+			} else {
+				authorizationPolicies[hash][index] = ap
+			}
+		} else {
+			hashTo, err := helpers.GetAuthorizationPolicyHash(ap)
+			if err != nil {
+				return nil, err
+			}
+
+			if authorizationPolicies[hashTo] == nil {
+				authorizationPolicies[hashTo] = make(map[string]*securityv1beta1.AuthorizationPolicy)
+			}
+
+			if _, ok := authorizationPolicies[hashTo][index]; ok {
+				authorizationPolicies[hashTo][index+":"+strconv.Itoa(i)] = ap
+			} else {
+				authorizationPolicies[hashTo][index] = ap
+			}
+		}
+	}
 
 	return authorizationPolicies, nil
 }
 
-func (r AuthorizationPolicyProcessor) getObjectChanges(desiredAps []*securityv1beta1.AuthorizationPolicy, actualAps []*securityv1beta1.AuthorizationPolicy) []*processing.ObjectChange {
+func (r AuthorizationPolicyProcessor) getObjectChanges(desiredAps map[string][]*securityv1beta1.AuthorizationPolicy, actualAps map[string]map[string]*securityv1beta1.AuthorizationPolicy) []*processing.ObjectChange {
 	var apObjectActionsToApply []*processing.ObjectChange
 
-	for _, ap := range desiredAps {
-		objectAction := processing.NewObjectCreateAction(ap)
-		apObjectActionsToApply = append(apObjectActionsToApply, objectAction)
+	for hashTo, toDesiredAPs := range desiredAps {
+		for _, ap := range toDesiredAPs {
+			// As both the order of Authorizations and APs is static we can update them according to array index
```

Comment:
This resolves the problem of having APIRule configuration like this one:
```yaml
- accessStrategies:
      - config:
          authentications:
          - issuer: https://kymagoattest.accounts400.ondemand.com
            jwksUri: https://kymagoattest.accounts400.ondemand.com/oauth2/certs
          authorizations:
          - audiences:
            - https://example.com
            requiredScopes:
            - read
          - requiredScopes:
            - write
        handler: jwt
```
The problem with this configuration is that if both of the scopes would change, we cannot be sure with the hashcode that update is executed on the AP corresponding to e.g. `read` scope, which may cause service downtime, as we would have to APs that are exposing the same path at the same time, and the AP for `write` would not be present